src/main/scala/Ch4.scala

```diff
@@ -0,0 +1,204 @@
+    val (aToString, aOrBToInt) = p3DAB
+    val aOrCToInt: Either[A, C] => Int = {
+      case Left(a) => aOrBToInt(Left(a))
+      case Right(c) => aOrBToInt(Right(f(c)))
+    }
+    (aToString, aOrCToInt)
+  }
+
+  // A => C
+  def p3ContraFmapAC[A, B, C](f: C => A): P3Data[A, B] => P3Data[C, B] = { p3DAB =>
+    val (aToString, aOrBToInt) = p3DAB
+    val cToString: C => String = c => aToString(f(c))
+    val cOrBToInt: Either[C, B] => Int = {
+      case Left(c) => aOrBToInt(Left(f(c)))
+      case Right(b) => aOrBToInt(Right(b))
+    }
+    (cToString, cOrBToInt)
+  }
+
+
+  // Problem 4
+  // Data[A] ≡ (1 + (A ⇒ String)) ⇒ (1 + (A ⇒ Int)) ⇒ Int
+  // Covariant
+  // The following will pass:
+  // final case class P4Data[A](d: Option[A ⇒ String] ⇒ Option[A ⇒ Int] ⇒ Int)
+  // import io.chymyst.ch._
+  // def p4Fmap[A, B](f: A => B): P4Data[A] => P4Data[B] = implement
+  // not pass: def p4ContraFmap[A, B](f: B => A): P4Data[A] => P4Data[B] = implement
+  type P4Data[A] = Option[A => String] => Option[A => Int] => Int
+  def p4Fmap[A, B](f: A => B): P4Data[A] => P4Data[B] = { p4DataA =>
+    // p4DataA: Option[A => String] => Option[A => Int] => Int
+
+    val p4DataB: Option[B ⇒ String] ⇒ Option[B ⇒ Int] ⇒ Int = { optBToString => optBToInt =>
+      // optBToString: Option[B => String]
+      // optBToInt: Option[B => Int]
+
+      // First produce Option[A => Int] => Int using p4DataA and optBToString
+      val optToIntA: Option[A => Int] => Int = optBToString match {
+        case Some(bToStr) => {
+          val aToStr: A => String = a => bToStr(f(a))
+          p4DataA(Some(aToStr))
+        }
+        case None => {
+          p4DataA(None)
+        }
+      }
+
+      // Second produce Int from optBToInt using optToIntA and optBToInt
+      val i2: Int = optBToInt match {
+        case Some(bToInt) => {
+          val aToInt: A => Int = a => bToInt(f(a))
+          optToIntA(Some(aToInt))
+        }
+        case None => optToIntA(None)
+      }
+      // Must return integer
+      i2
+    }
+    p4DataB
+  }
+
+
+  // Problem 5
+  // Data[B] ≡ (B + (Int ⇒ B)) × (B + (String ⇒ B))
+  // Covariant
+  type P5Data[B] = (BOrXToB[B, Int], BOrXToB[B, String])
+
+  // Sub-structure is covariant wrt B since it's not consumed
+  type BOrXToB[B, X] = Either[B, X => B]
+
+  def fmapBOrXToB[B, C, X](f: B => C): BOrXToB[B, X] => BOrXToB[C, X] = {
+    case Left(b) => Left(f(b))
+    case Right(xToB) => Right(x => f(xToB(x)))
+  }
+
+  def p5Fmap[B, C](f: B => C): P5Data[B] => P5Data[C] = { p5DB =>
+    val (bOrIntToB, bOrStrToB) = p5DB
+    (fmapBOrXToB(f)(bOrIntToB), fmapBOrXToB(f)(bOrStrToB))
+  }
+
+
+  // Problem 6
+  // Short notation:
+  // Data[A, B] ≡ ((A x B) x (B => Int)) + (A x B x Int) + ((String => A) x (B => A))
+  // A is never consumed, so it is covariant
+  // B behaves like a functor in case class `Re` but is consumed in the others,
+  // so it's invariant
+  sealed trait Coi[+A, B]
```

Comment:
You have taken the "worked example" (Coi = Pa + Re + Ci) instead of the data type defined in exercise 6 (Result = P + Q + R).

Reply:
Lol sorry about that. I now see that the solution to my submission was already there in your worked examples.